src/routes/Project/actions/TrackActionHandler.js

```diff
@@ -47,101 +82,168 @@ const ACTION_HANDLERS = {
-          ...state[newClip.track],
-          [newClip.startTime]: id,
-          [newClip.endTime]: id,
+        [oldClip.track]: {
+          ...state[oldClip.track],
           clips: {
-            ...state[newClip.track].clips,
+            ...state[oldClip.track].clips,
             [id]: newClip
           }
         }
       };
 
-      delete newTracks[newClip.track][oldClip.startTime];
-      delete newTracks[newClip.track][oldClip.endTime];
-      newTracks[newClip.track].edgeClip = findEdgeClip(newTracks[newClip.track]);
+      newTracks[oldClip.track].edgeClip = findEdgeClip(newTracks[oldClip.track].clips);
     }
 
     return newTracks;
+  },
+
+  [ProjectConstants.DELETE_CLIP]: (state, action) => {
+    const clip = action.payload;
+    let newClips = { ...state[clip.track].clips };
+    delete newClips[clip.id];
+
+    return {
+      ...state,
+      [clip.track]: {
+        ...state[clip.track],
+        clips: newClips,
+        edgeClip: findEdgeClip(newClips)
+      }
+    };
   }
 };
 
 export default ACTION_HANDLERS;
 
 /* HELPERS */
-function findEdgeClip(track) {
+function findEdgeClip(clips) {
   let max;
-  for (let time in track) {
+  let maxClip;
+  for (let clipId in clips) {
+    let clip = clips[clipId];
+    let time = clip.endTime;
     time = Number(time);
     if (typeof time === 'number' && !Number.isNaN(time)) {
-      if (!max || time > max) max = time;
+      if (!max || time > max) {
+        max = time;
+        maxClip = clip;
+      }
     }
   }
 
   if (!max) return null;
-  return track.clips[track[max]];
+  return maxClip;
+}
+
+// handle collision avoidance. not very efficient but should work.
+function adjustForCollisions(newClip, track, isMovingRight) {
+  const times = _(track.clips).keys()
+    .reduce((acc, currTrackId) => {
+      let currTrack = track.clips[currTrackId];
+      acc.push({ time: Number(currTrack.startTime), id: currTrackId });
+      acc.push({ time: Number(currTrack.endTime), id: currTrackId });
+      return acc;
+    }, [])
+    .sort(sortAscendingTimeEntries);
+
+  for (let { id } of times) {
+    let timeClip = track.clips[id];
+    if (timeClip.id === newClip.id) continue;
+    let clipLength = common.getClipLength(timeClip);
+
+    // for each clip, if the new clip ends after
+    // clip starts and the new clip starts before the clip ends
+    if (newClip.endTime > timeClip.startTime && newClip.startTime <= timeClip.endTime) {
+
+      // snap either to that clip's right or left,
+      // depending on direction of motion (if right, snap to left, etc)
+      if (isMovingRight) {
+        newClip.endTime = timeClip.startTime - 1;
+        newClip.startTime = timeClip.startTime - clipLength - 1;
+      } else {
+        newClip.endTime = timeClip.endTime + clipLength + 1;
+        newClip.startTime = timeClip.endTime + 1;
+      }
+
+      // and recurse
+      adjustForCollisions(newClip, track, isMovingRight);
+    }
+  }
+}
+
+function isCollision(newClip, track) {
```

Comment:
ya thats because there are cases where I need to know if theres a collision without having the clip repositioned. how were you thinking of keeping that functionality in one function. the best I can think of is using the isCollision function within the adjustForCollisions functions.